security-proxy/src/main/java/org/georchestra/security/LdapUserDetailsRequestHeaderProvider.java

```diff
@@ -39,163 +51,323 @@
+ * default headers.
+ * <p>
+ * Service name matches the ones assigned in {@code targets-mapping.properties}.
+ * For example, for service {@code analytics},
+ * {@code targets-mappings.properties} contains
+ * {@code analytics=http://analytics:8080/analytics/}, and
+ * {@code headers-mappings.properties} may contain
+ * {@code analytics.sec-firstname=givenName}.
+ * 
+ * 
  * @author jeichar
+ * @see SecurityRequestHeaderProvider
  */
 public class LdapUserDetailsRequestHeaderProvider extends HeaderProvider {
 
+    private static final String CACHED_USERNAME_KEY = "security-proxy-cached-username";
+
+    private static final String CACHED_HEADERS_KEY = "security-proxy-cached-attrs";
+
     protected static final Log logger = LogFactory
             .getLog(LdapUserDetailsRequestHeaderProvider.class.getPackage().getName());
 
-    private LdapUserSearch _userSearch;
-    private Map<String, String> _headerMapping;
-    private Pattern pattern;
-    private String orgSearchBaseDN;
+    private final FilterBasedLdapUserSearch _userSearch;
+    private final Pattern orgSeachMemberOfPattern;
+    private final String orgSearchBaseDN;
+
+    /**
+     * Header mappings that apply to all services (i.e. have no service prefix in
+     * header-mappings.properties), for example: ({@code sec-email=mail}
+     */
+    Map<String, String> defaultMappinsg = ImmutableMap.of();
+    /**
+     * Header mappings that apply to a specific target service, by service name,
+     * where service name matches the ones assigned in
+     * {@code targets-mapping.properties}. For example, for service
+     * {@code analytics}, {@code targets-mappings.properties} contains
+     * {@code analytics=http://analytics:8080/analytics/}, and
+     * {@code headers-mappings.properties} may contain
+     * {@code analytics.sec-firstname=givenName}.
+     */
+    Map<String, Map<String, String>> perServiceMappings = ImmutableMap.of();
 
     @Autowired
-    private LdapTemplate ldapTemplate;
+    @VisibleForTesting
+    LdapTemplate ldapTemplate;
 
     @Autowired
     private GeorchestraConfiguration georchestraConfiguration;
 
-    public LdapUserDetailsRequestHeaderProvider(LdapUserSearch userSearch, String orgSearchBaseDN,
-            Map<String, String> headerMapping) {
+    public LdapUserDetailsRequestHeaderProvider(FilterBasedLdapUserSearch userSearch, String orgSearchBaseDN) {
         Assert.notNull(userSearch, "userSearch must not be null");
-        Assert.notNull(headerMapping, "headerMapping must not be null");
+        // Q: is it ok for orgSearchBaseDN to be null?
         this._userSearch = userSearch;
-        this._headerMapping = headerMapping;
         this.orgSearchBaseDN = orgSearchBaseDN;
-
-        this.pattern = Pattern.compile("([^=,]+)=([^=,]+)," + orgSearchBaseDN + ".*");
+        this.orgSeachMemberOfPattern = Pattern.compile("([^=,]+)=([^=,]+)," + orgSearchBaseDN + ".*");
```

Comment:
typo orgSearchMemberOfPattern (with a 'r' to search)

Reply:
Fixed. 2e283fc9e5a94395359a31956b871cab3b6999d6